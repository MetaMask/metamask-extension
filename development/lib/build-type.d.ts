import { Infer, Struct } from 'superstruct';

// TODO: Replace `any` with type
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Unique<Element extends Struct<any>> = (
  struct: Struct<Infer<Element>[], Infer<Element>>,
  eq?: (a: Infer<Element>, b: Infer<Element>) => boolean,
) => Struct<Infer<Element>[], Infer<Element>>;
