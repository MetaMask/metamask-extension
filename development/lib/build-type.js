const fs = require('fs');
const { AssertionError } = require('assert');
const path = require('path');
const {
  object,
  string,
  record,
  optional,
  array,
  refine,
  any,
  boolean,
  coerce,
  union,
  number,
  unknown,
  validate,
  nullable,
  never,
  literal,
} = require('superstruct');
const yaml = require('yaml');
const { uniqWith } = require('lodash');

const BUILDS_YML_PATH = path.resolve('./builds.yml');

/**
 * @type {import('superstruct').Infer<typeof BuildTypesStruct> | null}
 */
let cachedBuildTypes = null;

/**
 * Ensures that the array item contains only elements that are distinct from each other
 *
 * @template {Struct<any>} Element
 * @type {import('./build-type').Unique<Element>}
 */
const unique = (struct, eq) =>
  refine(struct, 'unique', (value) => {
    if (uniqWith(value, eq).length === value.length) {
      return true;
    }
    return 'Array contains duplicated values';
  });

const EnvDefinitionStruct = coerce(
  object({ key: string(), value: unknown() }),
  refine(record(string(), any()), 'Env variable declaration', (value) => {
    if (Object.keys(value).length !== 1) {
      return 'Declaration should have only one property, the name';
    }
    return true;
  }),
  (value) => ({ key: Object.keys(value)[0], value: Object.values(value)[0] }),
);

const EnvArrayStruct = unique(
  array(union([string(), EnvDefinitionStruct])),
  (a, b) => {
    const keyA = typeof a === 'string' ? a : a.key;
    const keyB = typeof b === 'string' ? b : b.key;
    return keyA === keyB;
  },
);

/**
 * Ensures a number is within a given range
 *
 * @param {number} min
 * @param {number} max
 */
const isInRange = (min, max) => {
  /**
   *
   * @param {number} value
   * @returns boolean
   */
  function check(value) {
    return value >= min && value <= max;
  }
  return refine(number(), 'range', check);
};

const BuildTypeStruct = object({
  id: isInRange(10, 64),
  features: optional(unique(array(string()))),
  env: optional(EnvArrayStruct),
  isPrerelease: optional(boolean()),
  manifestOverrides: union([string(), literal(false)]),
  buildNameOverride: union([string(), literal(false)]),
});

const CopyAssetStruct = object({ src: string(), dest: string() });
const ExclusiveIncludeAssetStruct = coerce(
  object({ exclusiveInclude: string() }),
  string(),
  (exclusiveInclude) => ({ exclusiveInclude }),
);
const AssetStruct = union([CopyAssetStruct, ExclusiveIncludeAssetStruct]);

const FeatureStruct = object({
  env: optional(EnvArrayStruct),
  // TODO(ritave): Check if the paths exist
  assets: optional(array(AssetStruct)),
});

const FeaturesStruct = refine(
  record(
    string(),
    coerce(FeatureStruct, nullable(never()), () => ({})),
  ),
  'feature definitions',
  function* (value) {
    let isValid = true;

    const definitions = new Set();

    for (const feature of Object.values(value)) {
      for (const env of feature?.env ?? []) {
        if (typeof env !== 'string') {
          if (definitions.has(env.key)) {
            isValid = false;
            yield `Multiple defined features have a definition of "${env}" env variable, resulting in a conflict`;
          }
          definitions.add(env.key);
        }
      }
    }
    return isValid;
  },
);

const BuildTypesStruct = refine(
  object({
    default: string(),
    buildTypes: record(string(), BuildTypeStruct),
    features: FeaturesStruct,
    env: EnvArrayStruct,
  }),
  'BuildTypes',
  (value) => {
    if (!Object.keys(value.buildTypes).includes(value.default)) {
      return `Default build type "${value.default}" does not exist in builds declarations`;
    }
    return true;
  },
);

/**
 * Loads definitions of build type and what they are composed of.
 *
 * @returns {import('superstruct').Infer<typeof BuildTypesStruct>}
 */
function loadBuildTypesConfig() {
  if (cachedBuildTypes !== null) {
    return cachedBuildTypes;
  }
  const buildsData = yaml.parse(fs.readFileSync(BUILDS_YML_PATH, 'utf8'));
  const [err, result] = validate(buildsData, BuildTypesStruct, {
    coerce: true,
  });
  if (err !== undefined) {
    throw new AssertionError({
      message: constructFailureMessage(err),
    });
  }
  cachedBuildTypes = result;
  return cachedBuildTypes;
}

/**
 * Creates a user readable error message about parse failure.
 *
 * @param {import('superstruct').StructError} structError
 * @returns {string}
 */
function constructFailureMessage(structError) {
  return `Failed to parse builds.yml
  -> ${structError
    .failures()
    .map(
      (failure) =>
        `${failure.message} (${BUILDS_YML_PATH}:.${failure.path.join('/')})`,
    )
    .join('\n  -> ')}
`;
}

module.exports = { loadBuildTypesConfig };
