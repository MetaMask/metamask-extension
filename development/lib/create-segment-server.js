const http = require('http');

/**
 * This is the default error handler to be used by this mock segment server.
 * It will print the error to the console and exit the process.
 *
 * @param {Error} error - The server error
 */
function defaultOnError(error) {
  console.log(error);
  process.exit(1);
}

/**
 * @typedef {import('http').IncomingMessage} IncomingMessage
 * @typedef {import('http').ServerResponse} ServerResponse
 */

/**
 * This function handles requests for the mock Segment server
 *
 * @typedef {(request: IncomingMessage, response: ServerResponse, metricEvents: Array<object>) => void} MockSegmentRequestHandler
 */

/**
 * Creates a HTTP server that acts as a fake version of the Segment API.
 * The bevahiour is rudimentary at the moment - it returns HTTP 200 in response
 * to every request. The only function this serves is to spy on requests sent to
 * this server, and to parse the request payloads as Segment batch events.
 *
 * @param {MockSegmentRequestHandler} onRequest - A callback for each request the server receives.
 * @param {(error: Error) => void} [onError] - A callback for server error events
 */
function createSegmentServer(onRequest, onError = defaultOnError) {
  const server = http.createServer(async (request, response) => {
    const chunks = [];

    request.on('data', (chunk) => {
      chunks.push(chunk);
    });

    await new Promise((resolve) => {
      request.on('end', () => {
        resolve();
      });
    });

    // respond to preflight request
    if (request.method === 'OPTIONS') {
      response.setHeader('Access-Control-Allow-Origin', '*');
      response.setHeader('Access-Control-Allow-Methods', '*');
      response.setHeader('Access-Control-Allow-Headers', '*');
      response.statusCode = 200;
      response.end();
      return;
    }

    let metricEvents = [];
    if (chunks.length) {
      const body = Buffer.concat(chunks).toString();
      const segmentPayload = JSON.parse(body);
      metricEvents = segmentPayload.batch;
    }

    onRequest(request, response, metricEvents);
  });

  server.on('error', onError);

  return {
    start: async (port) => {
      await new Promise((resolve, reject) => {
        server.listen(port, (error) => {
          if (error) {
            return reject(error);
          }
          return resolve();
        });
      });
    },
    stop: async () => {
      await new Promise((resolve, reject) => {
        server.close((error) => {
          if (error) {
            return reject(error);
          }
          return resolve();
        });
        // We need to close all connections to stop the server quickly
        // Otherwise it takes a few seconds for it to close
        server.closeAllConnections();
      });
    },
  };
}

module.exports = { createSegmentServer };
