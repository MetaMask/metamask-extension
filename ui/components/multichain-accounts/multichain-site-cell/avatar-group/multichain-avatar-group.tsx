import React from 'react';
import classnames from 'classnames';
import {
  AlignItems,
  BorderRadius,
  Display,
} from '../../../../helpers/constants/design-system';
import {
  AvatarAccount,
  AvatarAccountSize,
  AvatarAccountVariant,
  Box,
} from '../../../component-library';

type MultichainAccountAvatarGroupProps = {
  className?: string;
  limit?: number;
  members: { avatarValue: string; symbol?: string }[];
};

export const MultichainAccountAvatarGroup: React.FC<
  MultichainAccountAvatarGroupProps
> = ({ className = '', limit = 4, members = [] }): JSX.Element => {
  const visibleMembers = members.slice(0, limit).reverse();

  return (
    <Box
      alignItems={AlignItems.center}
      display={Display.Flex}
      className={classnames('multichain-avatar-group', className)}
      data-testid="avatar-group"
      gap={1}
    >
      <Box display={Display.Flex} alignItems={AlignItems.center} gap={1}>
        {visibleMembers.map((member, i) => {
          return (
            <Box borderRadius={BorderRadius.full} key={i}>
              <AvatarAccount
                data-testid={`avatar-account-${i}`}
                size={AvatarAccountSize.Xs}
                address={member.avatarValue}
                // TODO: Switch to maskicon once it is available
                variant={AvatarAccountVariant.Blockies}
              />
            </Box>
          );
        })}
      </Box>
    </Box>
  );
};
