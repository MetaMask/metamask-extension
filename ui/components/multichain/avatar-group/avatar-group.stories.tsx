import React from 'react';
import { StoryFn, Meta } from '@storybook/react';
import { AvatarGroup } from '.';

export default {
  title: 'Components/Multichain/AvatarGroup',
  component: AvatarGroup,
  argTypes: {
    limit: {
      control: 'number',
    },
    members: {
      control: 'object',
    },
  },
  args: {
    members: [
      { symbol: 'ETH', avatarValue: './images/eth_logo.svg' },
      { symbol: 'MATIC', avatarValue: './images/pol-token.svg' },
      { symbol: 'OP', avatarValue: './images/optimism.svg' },
      { symbol: 'AVAX', avatarValue: './images/avax-token.svg' },
    ],
    limit: 4,
  },
} as Meta<typeof AvatarGroup>;

const Template = (args) => <AvatarGroup {...args} />;

export const DefaultStory = Template.bind({});
DefaultStory.storyName = 'Default';

export const WithTag: StoryFn<typeof AvatarGroup> = (args) => (
  <AvatarGroup {...args} />
);
WithTag.args = {
  members: [
    { symbol: 'ETH', avatarValue: './images/eth_logo.svg' },
    { symbol: 'MATIC', avatarValue: './images/pol-token.svg' },
    { symbol: 'OP', avatarValue: './images/optimism.svg' },
    { symbol: 'AVAX', avatarValue: './images/avax-token.svg' },
    { symbol: 'PALM', avatarValue: './images/palm.svg' },
  ],
  limit: 2,
};

export const TokenWithOutSrc: StoryFn<typeof AvatarGroup> = (args) => (
  <AvatarGroup {...args} />
);
TokenWithOutSrc.args = {
  members: [
    { symbol: 'ETH', avatarValue: '' },
    { symbol: 'MATIC', avatarValue: '' },
    { symbol: 'OP', avatarValue: '' },
    { symbol: 'AVAX', avatarValue: '' },
  ],
  limit: 2,
};
