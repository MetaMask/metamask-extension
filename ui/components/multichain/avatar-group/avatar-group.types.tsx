import { BorderColor } from '../../../helpers/constants/design-system';
import { AvatarTokenSize } from '../../component-library';
import type { StyleUtilityProps } from '../../component-library/box';

export type AvatarGroupProps = StyleUtilityProps & {
  /** * Additional class name for the AvatarGroup component */
  className?: string;
  /** * Limit to show only a certain number of tokens and extras in Text */
  limit: number;
  /** * Type of Avatar: Token/Account */
  avatarType?: AvatarType;
  /** * List of Avatar Tokens */
  members: {
    /** * Image of Avatar Token */
    avatarValue: string;
    /** * Symbol of Avatar Token */
    symbol?: string;
  }[];
  /** * Size of Avatar Tokens. For AvatarGroup we are considering AvatarTokenSize.Xs, AvatarTokenSize.Sm, AvatarTokenSize.Md  */
  size?: AvatarTokenSize;
  /** * Border Color of Avatar Tokens */
  borderColor?: BorderColor;
  /** * Whether the tag should be displayed as separate text or within an overlay avatar */
  isTagOverlay?: boolean;
};

export enum AvatarType {
  TOKEN = 'TOKEN',
  ACCOUNT = 'ACCOUNT',
  NETWORK = 'NETWORK',
}
