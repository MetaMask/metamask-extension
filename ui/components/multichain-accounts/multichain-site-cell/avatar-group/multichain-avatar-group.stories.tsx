import React from 'react';
import { StoryFn, Meta } from '@storybook/react';
import { MultichainAccountAvatarGroup } from './multichain-avatar-group';

// Reusable avatar data for consistency across stories
const mockAvatar1 = {
  avatarValue: '0x0dcd5d886577d5081b0c52e242ef29e70be3e7bc',
  symbol: 'ETH',
};

const mockAvatar2 = {
  avatarValue: '0x123456789abcdef0123456789abcdef012345678',
  symbol: 'MATIC',
};

const mockAvatar3 = {
  avatarValue: '0xabcdef1234567890abcdef1234567890abcdef12',
  symbol: 'BNB',
};

const mockAvatar4 = {
  avatarValue: '0x987654321fedcba0987654321fedcba0987654321',
  symbol: 'USDC',
};

const mockAvatar5 = {
  avatarValue: '0x5555555555555555555555555555555555555555',
  symbol: 'DAI',
};

export default {
  title: 'Components/MultichainAccounts/MultichainAccountAvatarGroup',
  component: MultichainAccountAvatarGroup,
  parameters: {
    docs: {
      description: {
        component:
          'A component that displays a group of account avatars with configurable limits and styling',
      },
    },
  },
  argTypes: {
    className: {
      control: 'text',
      description: 'Additional CSS class name for styling',
    },
    limit: {
      control: { type: 'number', min: 1, max: 10 },
      description: 'Maximum number of avatars to display',
    },
    members: {
      control: 'object',
      description:
        'Array of member objects with avatarValue and optional symbol',
    },
  },
} as Meta<typeof MultichainAccountAvatarGroup>;

const Template: StoryFn<typeof MultichainAccountAvatarGroup> = (args) => (
  <div
    style={{
      width: '300px',
      height: '100px',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      border: '1px solid #ccc',
      borderRadius: '8px',
      padding: '16px',
      backgroundColor: '#f5f5f5',
    }}
  >
    <MultichainAccountAvatarGroup {...args} />
  </div>
);

export const Default = Template.bind({});
Default.args = {
  members: [mockAvatar1, mockAvatar2, mockAvatar3],
};

export const WithLimit = Template.bind({});
WithLimit.args = {
  members: [mockAvatar1, mockAvatar2, mockAvatar3, mockAvatar4, mockAvatar5],
  limit: 3,
};

export const ManyMembers = Template.bind({});
ManyMembers.args = {
  members: Array.from({ length: 10 }, (_, index) => ({
    avatarValue: `0x${index.toString().padStart(40, '0')}`,
    symbol: `TOKEN${index}`,
  })),
  limit: 4,
};

export const SingleMember = Template.bind({});
SingleMember.args = {
  members: [mockAvatar1],
};

export const NoSymbols = Template.bind({});
NoSymbols.args = {
  members: [
    { avatarValue: mockAvatar1.avatarValue },
    { avatarValue: mockAvatar2.avatarValue },
    { avatarValue: mockAvatar3.avatarValue },
  ],
};

export const EmptyGroup = Template.bind({});
EmptyGroup.args = {
  members: [],
};

export const WithCustomClassName = Template.bind({});
WithCustomClassName.args = {
  members: [mockAvatar1, mockAvatar2, mockAvatar3],
  className: 'custom-avatar-group-class',
};

export const HighLimit = Template.bind({});
HighLimit.args = {
  members: [mockAvatar1, mockAvatar2, mockAvatar3],
  limit: 10,
};
