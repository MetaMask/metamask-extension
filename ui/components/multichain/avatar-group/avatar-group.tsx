import * as React from 'react';
import classnames from 'classnames';
import { useSelector } from 'react-redux';
import { getUseBlockie } from '../../../selectors';
import {
  Text,
  Box,
  AvatarToken,
  AvatarTokenSize,
  AvatarAccount,
  AvatarAccountSize,
  AvatarAccountVariant,
  AvatarNetwork,
  AvatarNetworkSize,
  AvatarBase,
  AvatarBaseSize,
} from '../../component-library';
import {
  AlignItems,
  BackgroundColor,
  BorderColor,
  BorderRadius,
  Display,
  TextColor,
  TextVariant,
} from '../../../helpers/constants/design-system';
import { AvatarGroupProps, AvatarType } from './avatar-group.types';

export const AvatarGroup: React.FC<AvatarGroupProps> = ({
  className = '',
  limit = 4,
  members = [],
  size = AvatarTokenSize.Xs,
  avatarType = AvatarType.TOKEN,
  borderColor,
  isTagOverlay = false,
}): JSX.Element => {
  const membersCount = members.length;
  const visibleMembers = members.slice(0, limit).reverse();
  const showTag = membersCount > limit;
  const useBlockie = useSelector(getUseBlockie);

  let marginLeftValue = '';
  if (AvatarTokenSize.Xs) {
    marginLeftValue = '-8px';
  } else if (AvatarTokenSize.Sm) {
    marginLeftValue = '-12px';
  } else {
    marginLeftValue = '-16px';
  }
  const tagValue = `+${(membersCount - limit).toLocaleString()}`;
  return (
    <Box
      alignItems={AlignItems.center}
      display={Display.Flex}
      className={classnames('multichain-avatar-group', className)}
      data-testid="avatar-group"
      gap={1}
    >
      <Box display={Display.Flex} alignItems={AlignItems.center}>
        {visibleMembers.map((member, i) => {
          return (
            <Box
              borderRadius={BorderRadius.full}
              key={i}
              style={{ marginLeft: i === 0 ? '0' : marginLeftValue }}
            >
              {avatarType === AvatarType.TOKEN && (
                <AvatarToken
                  src={member.avatarValue}
                  name={member.symbol}
                  size={size}
                  borderColor={borderColor}
                />
              )}
              {avatarType === AvatarType.ACCOUNT && (
                <AvatarAccount
                  size={AvatarAccountSize.Xs}
                  address={member.avatarValue}
                  variant={
                    useBlockie
                      ? AvatarAccountVariant.Blockies
                      : AvatarAccountVariant.Jazzicon
                  }
                  borderColor={borderColor}
                />
              )}
              {avatarType === AvatarType.NETWORK && (
                <AvatarNetwork
                  src={member.avatarValue}
                  name={member.symbol ?? ''}
                  size={AvatarNetworkSize.Xs}
                />
              )}
            </Box>
          );
        })}
        {showTag && isTagOverlay && (
          <AvatarBase
            backgroundColor={BackgroundColor.overlayAlternative}
            style={{ marginLeft: marginLeftValue, fontSize: 8 }}
            size={AvatarBaseSize.Xs}
            borderColor={BorderColor.backgroundDefault}
            borderRadius={BorderRadius.MD}
            color={TextColor.overlayInverse}
          >
            {tagValue}
          </AvatarBase>
        )}
      </Box>
      {showTag && !isTagOverlay ? (
        <Box>
          <Text variant={TextVariant.bodySm} color={TextColor.textAlternative}>
            {tagValue}
          </Text>
        </Box>
      ) : null}
    </Box>
  );
};
