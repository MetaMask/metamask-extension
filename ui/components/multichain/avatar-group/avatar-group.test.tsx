/* eslint-disable jest/require-top-level-describe */
import { screen } from '@testing-library/react';
import React from 'react';
import configureMockStore from 'redux-mock-store';
import mockState from '../../../../test/data/mock-state.json';
import { renderWithProvider } from '../../../../test/lib/render-helpers';

import { AvatarGroup } from './avatar-group';

const members = [
  { symbol: 'ETH', avatarValue: './images/eth_logo.svg' },
  { symbol: 'MATIC', avatarValue: './images/pol-token.svg' },
  { symbol: 'OP', avatarValue: './images/optimism.svg' },
  { symbol: 'AVAX', avatarValue: './images/avax-token.svg' },
  { symbol: 'PALM', avatarValue: './images/palm.svg' },
];

describe('AvatarGroup', () => {
  const mockStore = configureMockStore()(mockState);

  it('should render AvatarGroup component', () => {
    const { getByTestId, container } = renderWithProvider(
      <AvatarGroup members={members} limit={4} />,
      mockStore,
    );
    expect(getByTestId('avatar-group')).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  it('should render the tag +1 if members has a length greater than limit', () => {
    renderWithProvider(<AvatarGroup members={members} limit={4} />, mockStore);

    expect(screen.getByText('+1')).toBeDefined();
  });

  it('should not render the tag if members has a length less than or equal to limit', () => {
    const { queryByText } = renderWithProvider(
      <AvatarGroup members={members} limit={5} />,
      mockStore,
    );
    expect(queryByText('+1')).not.toBeInTheDocument();
  });
});
