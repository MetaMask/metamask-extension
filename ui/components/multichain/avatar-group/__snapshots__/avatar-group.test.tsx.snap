// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarGroup should render AvatarGroup component 1`] = `
<div>
  <div
    class="mm-box multichain-avatar-group mm-box--display-flex mm-box--gap-1 mm-box--align-items-center"
    data-testid="avatar-group"
  >
    <div
      class="mm-box mm-box--display-flex"
    >
      <div
        class="mm-box mm-box--rounded-full"
        style="margin-left: 0px;"
      >
        <div
          class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-token mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full"
        >
          <img
            alt="AVAX logo"
            class="mm-avatar-token__token-image"
            src="./images/avax-token.svg"
          />
        </div>
      </div>
      <div
        class="mm-box mm-box--rounded-full"
        style="margin-left: -8px;"
      >
        <div
          class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-token mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full"
        >
          <img
            alt="OP logo"
            class="mm-avatar-token__token-image"
            src="./images/optimism.svg"
          />
        </div>
      </div>
      <div
        class="mm-box mm-box--rounded-full"
        style="margin-left: -8px;"
      >
        <div
          class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-token mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full"
        >
          <img
            alt="MATIC logo"
            class="mm-avatar-token__token-image"
            src="./images/pol-token.svg"
          />
        </div>
      </div>
      <div
        class="mm-box mm-box--rounded-full"
        style="margin-left: -8px;"
      >
        <div
          class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-token mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full"
        >
          <img
            alt="ETH logo"
            class="mm-avatar-token__token-image"
            src="./images/eth_logo.svg"
          />
        </div>
      </div>
    </div>
    <div
      class="mm-box"
    >
      <p
        class="mm-box mm-text mm-text--body-sm mm-box--color-text-alternative"
      >
        +1
      </p>
    </div>
  </div>
</div>
`;
